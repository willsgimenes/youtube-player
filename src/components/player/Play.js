import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'

import { togglePlay } from '../../actions'

function mapStateToProps ({ player }) {
  return { playing: player.playing }
}

function mapDispatchToProps (dispatch) {
  return bindActionCreators({ togglePlay }, dispatch)
}

class Play extends Component {
  constructor (props) {
    super(props)

    this.handleClick = this.handleClick.bind(this)
  }

  handleClick () {
    const { playing } = this.props
    this.props.togglePlay(!playing)
  }

  render () {
    const { playing } = this.props

    return (
      <button
        onClick={() => this.handleClick()}>
        { playing
          ? <i
            className='fa fa-pause'
            aria-hidden='true' />
          : <i
            className='fa fa-play'
            aria-hidden='true' />
        }
      </button>
    )
  }
}

Play.propTypes = {
  togglePlay: PropTypes.func.isRequired,
  playing: PropTypes.bool.isRequired
}

export default connect(mapStateToProps, mapDispatchToProps)(Play)
