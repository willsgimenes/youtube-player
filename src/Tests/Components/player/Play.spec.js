/* global describe */
/* global beforeEach */
/* global it */
/* global expect */

import React from 'react'
import Enzyme, { mount, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import Play from '../../../components/player/Play'

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

Enzyme.configure({ adapter: new Adapter() })

describe('Play button', () => {
  let wrapper, store, initialState

  beforeEach(() => {
    initialState = {
      player: {
        playing: false
      }
    }
    store = mockStore(initialState)
    wrapper = shallow(<Play store={store} />)
  })

  it('should have the playing props equals false', () => {
    expect(wrapper.props().playing).toBe(false)
  })

  it('should render properly', () => {
    const newInitialState = {
      player: {
        playing: false
      }
    }

    const newStore = mockStore(newInitialState)
    const newWrapper = mount(<Play store={newStore} />)
    expect(newWrapper.find('button')).toHaveLength(1)
  })

  it('should have the playing props equals true', () => {
    initialState = {
      player: {
        playing: true
      }
    }
    store = mockStore(initialState)
    wrapper = shallow(<Play store={store} />)

    expect(wrapper.props().playing).toBe(true)
  })

  it('should show pause icon when state.playing === true', () => {
    initialState = {
      player: {
        playing: true
      }
    }
    store = mockStore(initialState)
    wrapper = shallow(<Play store={store} />).dive()
    expect(wrapper.find('i').hasClass('fa fa-pause')).toBe(true)
    expect(wrapper).toMatchSnapshot()
  })

  it('should show play icon when state.playing === false', () => {
    initialState = {
      player: {
        playing: false
      }
    }
    store = mockStore(initialState)
    wrapper = shallow(<Play store={store} />).dive()
    expect(wrapper.find('i').hasClass('fa fa-play')).toBe(true)
    expect(wrapper).toMatchSnapshot()
  })

  it('should simulate click action', () => {
    wrapper.dive().simulate('click')
    const actions = wrapper.props().store.getActions()
    expect(actions).toEqual([{ payload: true, type: 'PLAY_UPDATE' }])
  })
})
