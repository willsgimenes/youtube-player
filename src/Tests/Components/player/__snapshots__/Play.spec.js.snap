// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Play button should show pause icon when state.playing === true 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Play
    playing={true}
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
    storeSubscription={
        Subscription {
            "listeners": Object {
              "clear": [Function],
              "get": [Function],
              "notify": [Function],
              "subscribe": [Function],
            },
            "onStateChange": [Function],
            "parentSub": undefined,
            "store": Object {
              "clearActions": [Function],
              "dispatch": [Function],
              "getActions": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
            },
            "unsubscribe": [Function],
          }
    }
    togglePlay={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <i
        aria-hidden="true"
        className="fa fa-pause"
/>,
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "aria-hidden": "true",
        "className": "fa fa-pause",
      },
      "ref": null,
      "rendered": null,
      "type": "i",
    },
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <i
          aria-hidden="true"
          className="fa fa-pause"
/>,
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "aria-hidden": "true",
          "className": "fa fa-pause",
        },
        "ref": null,
        "rendered": null,
        "type": "i",
      },
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "storeSubscription": undefined,
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`Play button should show play icon when state.playing === false 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Play
    playing={false}
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
    storeSubscription={
        Subscription {
            "listeners": Object {
              "clear": [Function],
              "get": [Function],
              "notify": [Function],
              "subscribe": [Function],
            },
            "onStateChange": [Function],
            "parentSub": undefined,
            "store": Object {
              "clearActions": [Function],
              "dispatch": [Function],
              "getActions": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
            },
            "unsubscribe": [Function],
          }
    }
    togglePlay={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <i
        aria-hidden="true"
        className="fa fa-play"
/>,
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "aria-hidden": "true",
        "className": "fa fa-play",
      },
      "ref": null,
      "rendered": null,
      "type": "i",
    },
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <i
          aria-hidden="true"
          className="fa fa-play"
/>,
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "aria-hidden": "true",
          "className": "fa fa-play",
        },
        "ref": null,
        "rendered": null,
        "type": "i",
      },
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "storeSubscription": undefined,
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
